src/core/load-dropin.c

```diff
@@ -19,51 +19,245 @@
+                        return r;
+        }
+
+        r = set_ensure_allocated(&u->deferred_dropin_dependencies, &dropin_hash_ops);
+        if (r < 0)
+                return r;
 
-        r = unit_add_dependency_by_name(u, dependency, entry, filepath, true);
+        r = set_ensure_allocated(&other->deferred_dropin_units, NULL);
         if (r < 0)
-                log_error_errno(r, "Cannot add dependency %s to %s, ignoring: %m", entry, u->id);
+                return r;
+
+        ds = new0(DropinSymlink, 1);
+        if (!ds)
+                return -ENOMEM;
 
+        ds->unit = other;
+        ds->depth = data->depth;
+        ds->dependency = dependency;
+        ds->is_mask = symlink_to_devnull(filepath);
+
+        r = set_put(u->deferred_dropin_dependencies, ds);
+        if (r <= 0)
+                /* r == 0 means we already registered the same dependency
+                 * for 'u' but with a higher priority (lower depth). */
+                return r;
+
+        r = set_put(other->deferred_dropin_units, u);
+        if (r < 0) {
+                set_remove(u->deferred_dropin_dependencies, ds);
+                return r;
+        }
+
+        ds = NULL;
         return 0;
 }
 
 int unit_load_dropin(Unit *u) {
+        DependencyConsumerData data = { .unit = u, .depth = 0 };
         _cleanup_strv_free_ char **l = NULL;
         Iterator i;
-        char *t, **f;
+        char **f, **p;
         int r;
 
         assert(u);
 
         /* Load dependencies from supplementary drop-in directories */
 
-        SET_FOREACH(t, u->names, i) {
-                char **p;
+        STRV_FOREACH(p, u->manager->lookup_paths.search_path) {
+                char *t;
 
-                STRV_FOREACH(p, u->manager->lookup_paths.search_path) {
-                        unit_file_process_dir(u->manager->unit_path_cache, *p, t, ".wants", UNIT_WANTS,
-                                              add_dependency_consumer, u, NULL);
-                        unit_file_process_dir(u->manager->unit_path_cache, *p, t, ".requires", UNIT_REQUIRES,
-                                              add_dependency_consumer, u, NULL);
+                SET_FOREACH(t, u->names, i) {
+                        unit_file_process_dir(u->manager->unit_path_cache, *p, t, ".wants",
+                                              UNIT_WANTS,
+                                              add_dependency_consumer, &data, NULL);
+                        unit_file_process_dir(u->manager->unit_path_cache, *p, t, ".requires",
+                                              UNIT_REQUIRES,
+                                              add_dependency_consumer, &data, NULL);
                 }
+                data.depth++;
         }
 
+        /* Try to load the deferred dependencies now just in
+         * case all dropins have been already loaded. */
+        r = unit_load_reserve_deferred_dropin_dependencies(u);
+        if (r < 0)
+                return r;
```

Comment:
when this fails, shouldn't we free the two Sets here?

I am a bit concerned about leaving them around, given that they might continue pointing to Unit objects that go away later...